File: ts/src/calc.ts
// A tiny arithmetic library, mirrored across every language in this fixture.

export function add(a: number, b: number): number {
  return a + b
}

export function sub(a: number, b: number): number {
  return a - b
}

export function mul(a: number, b: number): number {
  return a * b
}

export function divide(a: number, b: number): number {
  if (b === 0) {
    throw new Error('divide by zero')
  }
  return a / b
}

// The compound condition sits on one line and the tests exercise only some of
// its outcomes, so lcov reports the line as PARTIALLY covered (BRDA taken on a
// subset) — the fixture for partial-branch rendering.
export function clamp(n: number, lo: number, hi: number): number {
  if (n < lo || n > hi) {
    return n < lo ? lo : hi
  }
  return n
}

// The "negative" branch is intentionally left uncovered → keeps coverage < 100%.
export function classify(n: number): string {
  if (n < 0) {
    return 'negative'
  }
  if (n === 0) {
    return 'zero'
  }
  return 'positive'
}

// ---- Partial-branch fixtures ----
// Each function below puts several branch outcomes on one line and the tests
// exercise only a subset, so the line reports PARTIAL branch coverage.

// `&&` chain: tests never pass a valid port, so the right-hand conditions
// short-circuit away.
export function isValidPort(n: number, allowPrivileged = false): boolean {
  return Number.isInteger(n) && n > 0 && n < 65536 && (allowPrivileged || n >= 1024)
}

// Nullish coalescing: tests always pass a value, the fallback arm never runs.
export function withUnit(value: number, unit?: string): string {
  return `${value} ${unit ?? 'ms'}`
}

// Chained ternary: tests only ever reach the first arm.
export function grade(score: number): string {
  return score >= 90 ? 'A' : score >= 75 ? 'B' : score >= 50 ? 'C' : 'F'
}

// New in this PR — TEXTBOOK partial coverage: the test only ever passes n === 0,
// so the first operand is always true and `n === 999` is NEVER reached (the || short-
// circuits). The line runs (covered) but one branch outcome is missing → partial.
export function isSentinel(n: number): boolean {
  if (n === 0 || n === 999) {
    return true
  }
  return false
}

File: go/calc.go
// Package calc is a tiny arithmetic library used to exercise coverage tooling.
package calc

import "errors"

// ErrDivideByZero is returned by Divide when the divisor is zero.
var ErrDivideByZero = errors.New("divide by zero")

// Add returns a + b.
func Add(a, b int) int { return a + b }

// Mul returns a * b.
func Mul(a, b int) int { return a * b }

// Divide returns a / b, or ErrDivideByZero when b == 0.
func Divide(a, b int) (int, error) {
	if b == 0 {
		return 0, ErrDivideByZero
	}
	return a / b, nil
}

// Classify labels n as "negative", "zero", or "positive". The "negative" branch
// is intentionally left uncovered by tests so the fixture reports < 100% coverage.
func Classify(n int) string {
	switch {
	case n < 0:
		return "negative"
	case n == 0:
		return "zero"
	case n == 1:
		return "one"
	default:
		return "positive"
	}
}

// Pow returns base**exp for exp >= 0. New in this PR and covered by a test, so
// its lines render as "new covered" in the comparison.
func Pow(base, exp int) int {
	result := 1
	for i := 0; i < exp; i++ {
		result *= base
	}
	return result
}

// IsEven reports whether n is even. New in this PR and INTENTIONALLY left
// untested, so its lines render as "new uncovered" — the regression to fix.
func IsEven(n int) bool {
	return n%2 == 0
}

// IsSentinel reports whether n is a sentinel value. The test only ever passes 0,
// so `n == 0` is always true and `|| n == 999` is NEVER reached. NOTE: Go's
// coverage is line-level (no branch coverage), so this renders as covered — its
// TS and Java twins show the same line as PARTIAL.
func IsSentinel(n int) bool {
	if n == 0 || n == 999 {
		return true
	}
	return false
}
